Refactor Velux integration

pawlizio opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Config Flow added for Velux component
Tilt functionality added for Velux cover
Workaround for #34844 by providing a reboot_gateway service 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
velux:
  host: "192.168.xxx.xxx"
  password: !secret velux_password
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [X] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

pawlizio committed:
Velux refactor

epenet commented:
Two quick general comments:
- please use references in the config flow translations (#40578)
- to avoid breaking change, use an import routine to migrate from YAML to ConfigEntry (https://github.com/home-assistant/core/blob/9ee97cb213d659aa9b6149484c0a42522decba78/homeassistant/components/nut/sensor.py#L60-L67)

pawlizio committed:
several updates

pawlizio commented on homeassistant/components/velux/cover.py:
```diff
@@ -110,11 +140,31 @@ async def async_set_cover_position(self, **kwargs):
+        await self.node.close_orientation(wait_for_completion=False)
+
+    async def async_open_cover_tilt(self, **kwargs):
+        """Close the cover."""
```
Thanks, fixed now!

pawlizio commented on homeassistant/components/velux/cover.py:
```diff
@@ -110,11 +140,31 @@ async def async_set_cover_position(self, **kwargs):
+        await self.node.open_orientation(wait_for_completion=False)
+
+    async def async_stop_cover_tilt(self, **kwargs):
+        """Close the cover."""
```
Thanks, fixed now!

pawlizio committed:
Update en.json

pawlizio committed:
Update strings.json

pawlizio commented:
> Two quick general comments:
> 
> * please use references in the config flow translations (#40578)
> * to avoid breaking change, use an import routine to migrate from YAML to ConfigEntry (https://github.com/home-assistant/core/blob/9ee97cb213d659aa9b6149484c0a42522decba78/homeassistant/components/nut/sensor.py#L60-L67
>   )

Added referenced and corrected the Type of Change description, YAML to ConfigEntry was already included, so no breaking change.

epenet reviewed (commented):
I also suggest that you split this into 3 PR?
- one for the config flow (this will most likely take longest to review)
- one for adding tilt functionnality
- one for reboot gateway

Also, config_flow requires 100% test coverage so you will need to update `.coveragerc` with the list of files, and build some actual tests for the config_flow:
     ~~homeassistant/components/velux/*~~
     homeassistant/components/velux/cover.py
     homeassistant/components/velux/scene.py

_Note: I'm just a contributor not an official reviewer_

pawlizio committed:
Update en.json

epenet commented:
Ref  #42895

epenet commented:
Ref #42895

pawlizio committed:
Update config_flow.py

pawlizio commented on homeassistant/components/velux/translations/en.json:
```diff
@@ -0,0 +1,24 @@
+{
+    "config": {
+        "abort": {
+            "already_configured_device": "[%key:common::config_flow::error::already_configured_device%]",
```
I updated the translation file by using `python3 -m script.translations develop`. The order is changed automatically.

epenet commented:
@pawlizio did you see my note about test coverage? It is compulsory to have tests which cover the whole of the config_flow and until you have that the config_flow will never get merged into the core.

pawlizio commented:
Yes I saw, but not sure when I'll find the time to learn how to write that unit tests and finally to do it. I'm not a professional software developer, so it will take me some time which I do not have now. I'm fine with my custom component

Holen Sie sich Outlook für Android<https://aka.ms/ghei36>

________________________________
From: epenet <notifications@github.com>
Sent: Friday, November 13, 2020 10:34:02 AM
To: home-assistant/core <core@noreply.github.com>
Cc: Paul Daumlechner <paul.daumlechner@live.de>; Mention <mention@noreply.github.com>
Subject: Re: [home-assistant/core] Refactor Velux integration (#42773)


@pawlizio<https://github.com/pawlizio> did you see my note about test coverage? It is compulsory to have tests which cover the whole of the config_flow and until you have that the config_flow will never get merged into the core.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/home-assistant/core/pull/42773#issuecomment-726656945>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADS6GHULPLT434YM33IWJGLSPT4QVANCNFSM4TIAFW7A>.

mtdcr commented:
@pawlizio I'd recommend rebasing this once https://github.com/home-assistant/core/pull/43198 got merged, so the code stays in good shape.

@frenck Could you please remove the 'breaking-change' label?

pawlizio commented:
> @pawlizio I'd recommend rebasing this once #43198 got merged, so the code stays in good shape.

@mtdcr The reboot function is already included in this pull request: 
https://github.com/home-assistant/core/blob/a81e9fedd075e482224cb41a9675bed7c621aea4/homeassistant/components/velux/__init__.py#L45-L48

So if this would be integrated first, #43198 should be closed. If #43198 will be merged first it will cause a conflict which needs to be resolved. Can this be solved by a rebase?

I created #43198 because I assume such small pull request will be integrated much faster and I'm not sure if this one will ever be merged due to missing unit tests.

pawlizio committed:
Update const.py

pawlizio committed:
Update scene.py

pawlizio committed:
Update strings.json

pawlizio committed:
Update en.json

mtdcr commented:
> If #43198 will be merged first it will cause a conflict which needs to be resolved. Can this be solved by a rebase?

Yes, that's what I meant. Rebasing will probably cause a conflict, which you will have to solve manually, but the result will be a PR without the reboot part. If you'll face difficulties, I can try to assist. Just ping me by mail then (obi at saftware dot de).

The same procedure should be followed when splitting out tilt functionality. In the meantime, you can mark this PR as work in progress by prefixing the title with `WIP: `.

> I'm not sure if this one will ever be merged due to missing unit tests.

I think I'll be able to help with the unit tests later this year. I'm interested in the conversion to config flow, because it allows to start Home Assistant with the Velux gateway being offline (or still being busy rebooting).

dumpfheimer commented:
@pawlizio in en.json in line 24 there is an extra "," which caused an UI error for me.

All in all this is a HUGE improvement for me. I was working on exactly the same two issues of the velux component (startup+tilt) which makes this an early christmas present for me ;-)

One improvement I would like to propose: I tried unplugging the KLF and start hass (worked perfectly, event the entities showed up!). Then, I got "[pyvlx] Connecting to KLF 200." but the entities never got available. Is this to be expected?

If I can test anything please let me know

pawlizio committed:
Update en.json

pawlizio commented:
@frenck: Could you kindly remove the "breaking change" label. That modification does not break functionality, yaml import is considered.

pawlizio commented:
> @pawlizio in en.json in line 24 there is an extra "," which caused an UI error for me.

@dumpfheimer Thanks for the hint, I've corrected this.

> All in all this is a HUGE improvement for me. I was working on exactly the same two issues of the velux component (startup+tilt) which makes this an early christmas present for me ;-)

For the time beeing you can use my HACS custom component: https://github.com/pawlizio/my_velux

> One improvement I would like to propose: I tried unplugging the KLF and start hass (worked perfectly, event the entities showed up!). Then, I got "[pyvlx] Connecting to KLF 200." but the entities never got available. Is this to be expected?

"[pyvlx] Connecting to KLF 200." does not mean that the connection was established successfully. There is still a bug in the KLF200 firmware which blocks new connections after some time #34844

> If I can test anything please let me know

There are no unit tests until now in this PR and I'm not sure whether this PR will be merged without. So if you can contribute, that would be fine.

dumpfheimer commented:
For clarification:

I freshly started KLF but disconnected LAN. Then, I started HASS and waited for everything to come up. After that I plugged LAN to KLF and waited. The entities did not "activate" in HASS. I used the same test procedure while developing "my" asynchronous setup.

pawlizio committed:
Add ConfigEntryNotReady exception

pawlizio commented:
> For clarification:
> 
> I freshly started KLF but disconnected LAN. Then, I started HASS and waited for everything to come up. After that I plugged LAN to KLF and waited. The entities did not "activate" in HASS. I used the same test procedure while developing "my" asynchronous setup.

@dumpfheimer: Can you check if the latest commit solves that issue. Thx

pawlizio committed:
Update __init__.py

Removed pylint comment

dumpfheimer commented:
Works perfectly!

balloob reviewed (changes requested):
Please add tests for the config flow. They were automatically created when the config flow was generated.

balloob commented on homeassistant/components/velux/config_flow.py:
```diff
@@ -0,0 +1,91 @@
+            step_id="user", data_schema=data_schema, errors=errors
+        )
+
+    async def async_step_unignore(self, user_input):
```
No need to add this step.

balloob commented on homeassistant/components/velux/manifest.json:
```diff
@@ -2,6 +2,14 @@
+    "pyvlx==0.2.18"
+  ],
+  "zeroconf": [
+    "_http._tcp.local."
```
Update this to include a name/Mac filter: https://developers.home-assistant.io/docs/creating_integration_manifest#zeroconf

balloob commented on homeassistant/components/velux/cover.py:
```diff
@@ -69,13 +80,29 @@ def should_poll(self):
+        )
+        if self.current_cover_tilt_position is not None:
+            supported_features |= (
+                SUPPORT_OPEN_TILT
```
Tilt support should be split out to a new PR. 

can we generate the supported features inside the constructor?

pawlizio commented:
> Please add tests for the config flow. They were automatically created when the config flow was generated.
@balloob  I'm trying to identify what I've missed, as I could not find any information where tests are created automatically. Are you talking about automatically generated unit tests? If so, could you give me a hint where I can get more information about them.
Thanks!

dumpfheimer commented:
Not stale 😬

balloob commented:
This is the one that has been automatically generated https://github.com/home-assistant/core/blob/dev/script/scaffold/templates/config_flow/tests/test_config_flow.py

Please use that as a basis for your tests.

dumpfheimer commented:
🥷

dumpfheimer commented:
I would still very much appreciate seeing this PR in HA

pergolafabio commented:
Same here

GeertvanHorrik commented:
To not just +1 this, I am happy to financially contribute to the person implementing this, either via GitHub sponsorship or open collective. If more people follow, there should be enough incentive for someone to pick this up.

dumpfheimer commented:
Side note: I have been using this PR for some months now, without any issues.
I would still very much appreciate it being merged.

How can we support the process?

pergolafabio commented:
> Side note: I have been using this PR for some months now, without any issues.
> I would still very much appreciate it being merged.
> 
> How can we support the process?

Do you still need to reboot the gateway? On boot or on restart?

dumpfheimer commented:
Actually I have 3 automations that restart the gateway:
1. If HA comes up and the devices stay unavailable
2. If one of the devices becomes unavailable for 30 minutes
3. Timed reboot at 02:00 in the night with a "stop" command after some time to cause a reconnect
(I "reboot" the gateway by killing its power supply)

To clarify my statement:
The whole KLF200 thing is not working flawlessly - **but they all seem (to me) to be KLF200 Firmware issues.** 

To me this PR has introduced no new issues but improved the experience in HA

pergolafabio commented:
Can you post those automations?

dumpfheimer commented:
Here you go.
As mentioned, I reboot the gateway by cutting its power supply. That is said shell_command.restart_klf.

[velux_reboot_and_reconnect_at_night.txt](https://github.com/home-assistant/core/files/6731505/velux_reboot_and_reconnect_at_night.txt)
[velux_auto_reboot_when_unavailable.txt](https://github.com/home-assistant/core/files/6731506/velux_auto_reboot_when_unavailable.txt)
[velux_automation_hass_start.txt](https://github.com/home-assistant/core/files/6731507/velux_automation_hass_start.txt)

Larx commented:
For me it is always unpredictable if the Velux gateway is available after a restart of HA. Is there a reason why the pull request does not get merged?
Also the restart service of the integration does not seem to work for me, I always have to cut the power supply.

dumpfheimer commented:
There seems to be an issue when pyvlx disconnects before disabling the house monitor. But I believe these are issues that should be discussed over at pyvlx, not here in the PR

dumpfheimer commented:
@pawlizio I just saw that you have deleted the branch, could you kindly tell us if you are still on to this?

pergolafabio commented:
> Here you go.
> As mentioned, I reboot the gateway by cutting its power supply. That is said shell_command.restart_klf.
> 
> [velux_reboot_and_reconnect_at_night.txt](https://github.com/home-assistant/core/files/6731505/velux_reboot_and_reconnect_at_night.txt)
> [velux_auto_reboot_when_unavailable.txt](https://github.com/home-assistant/core/files/6731506/velux_auto_reboot_when_unavailable.txt)
> [velux_automation_hass_start.txt](https://github.com/home-assistant/core/files/6731507/velux_automation_hass_start.txt)

strange, i cant download those 3 files?
